### ISSUE TITLE
Clarification on how to use multiple constraint dependencies

### ISSUE THREAD

**Issue by venaturum** [python-poetry/poetry#6625]
Following on from[ this discussion](https://github.com/python-poetry/poetry/discussions/6613):

I am keen to understand exactly how to specify multiple constraint dependencies.  The [current example](https://python-poetry.org/docs/dependency-specification/#multiple-constraints-dependencies) in the docs is not broad enough to eliminate ambiguity.

I have put together an [imgur album](https://imgur.com/a/wqiEn4b) to make it clear different ways that multiple constraint dependencies might be able to be specified.  I'm not sure which of the possibilities shown in the album are valid or not.

I'm happy to shape this into an example in the docs once it is clarified.

**Comment by neersighted** [comment]
Example C looks to be incorrect as Python 3.8 is fully excluded.

I personally would choose none of your approaches here, though it may be useful to document them all as possibilities. Keeping in mind that each multiple constraint dependency entry essentially represents an alternate solution (and Poetry will prefer the more specific/newer versioned solution), I would instead write minimal constraints that specify the newest version possible on each group of Pythons:

e.g.

```toml
[[tool.poetry.dependencies.foo]]
python = ">=3.6,<3.7"
version = "^1.3"

[[tool.poetry.dependencies.foo]]
python = ">=3.7"
version = "^1.4"
```

The use of `>=3.7` is a matter of taste -- by default Poetry will specify `^3.7` (mostly meaningless in the face of the current Python versioning system/intent -- we just don't special case the Python version). You could specify an upper bound if you don't want your package to be installable on unknown future Python versions absent testing (or if you know that you don't work on 3.11 RCs currently, for instance). Your project-level Python constraint will of course have to agree.

The main takeaway here is reducing the search space for the solver by using minimal constraints while allowing for newer versions.

The drawback of the approach I show here is that `poetry update` will not capture a hypothetical foo 1.5 that is only compatible with 3.9+ -- in that case you could add a third constraint e.g. 

```toml
[[tool.poetry.dependencies.foo]]
python = ">=3.6,<3.7"
version = "^1.3"

[[tool.poetry.dependencies.foo]]
python = ">=3.7,<3.9"
version = "^1.4"

[[tool.poetry.dependencies.foo]]
python = ">=3.9"
version = "^1.4"
```

Ultimately there is no correct one-size fits all answer for every project -- instead of a discussion of the different approaches and why one might prefer one or the other is probably the more useful docs contribution. Focusing on real world examples like NumPy (I don't know any other pathological examples off the top of my head) is also very useful -- in order to keep the discussion grounded and something people can directly experiment with.

**Comment by venaturum** [comment]
> Example C looks to be incorrect as Python 3.8 is fully excluded.

Good pickup, the middle constraint should have been `{version = ">=1.1,<1.5", python = ">=3.7,<3.9"},`

> [[tool.poetry.dependencies.foo]]
> python = ">=3.6,<3.7"
> version = "^1.3"

> [[tool.poetry.dependencies.foo]]
> python = ">=3.7"
> version = "^1.4"

If these constraints are used, and the package is published to pypi, won't this preclude someone from using the package with Python=3.7 and foo=1.3?  If they have another dependency which requires foo<1.4 then presumably they won't be able to resolve the environment right?


Just to confirm, were all of the approaches defined in the imgur album correct in that they could all be actually used - and are equivalent with respect to that dependency matrix?  This was what was unclear to me as it depends on what Poetry does under the hood.

**Comment by neersighted** [comment]
Indeed, I am talking about the perspective of an application which likely wants to force newer versions. For a library, I would construct the constraints differently to instead prefer the oldest version/make them as wide as possible.

Regarding your imgur examples, A or C are good examples of what I would do for 'wide' library constraints. I would prefer C personally, as it's functionally the same with less work done by the solver/fewer constraints on built packages.

I would suggest against B because it will force the solver to do a lot more work (to examine more combinations of Python and versions) needlessly. I think that example would work (maybe try with NumPy?), but it will be a lot slower for no good reason, and is harder to reason about.

Typically we write constraints from the perspective of 'what versions do I want to allow' and not 'what Python versions does upstream allow,' which I would characterize the latter as. Obviously those two go hand in hand, but typically we use the 'vertical' direction in your matrix as it better reflects the problem space and requires less work from the solver.

**Comment by venaturum** [comment]
Hi Bjorn, when I try putting this in my pyproject file

```
[tool.poetry.dependencies.pandas]
python = ">=3.6.1,<3.10"
version = "^1"

[tool.poetry.dependencies.pandas]
python = "^3.10"
version = "^1.3.4"
```

I get the following error

`Invalid TOML file C:/.../pyproject.toml: Key "pandas" already exists.`

Did I misunderstand something?

**Comment by neersighted** [comment]
Hi @nymous -- that is invalid toml. For a table in a list, write `[[parent.list.table]]` or

```toml
[parent]
list = [
  {}
]
```
